pageserver/src/tenant/size.rs

```diff
@@ -88,13 +104,18 @@ pub(super) async fn gather_inputs(
             let gc_info = timeline.gc_info.read().unwrap();
 
             // similar to gc, but Timeline::get_latest_gc_cutoff_lsn() will not be updated before a
-            // new gc run, which we have no control over.
-            // maybe this should be moved to gc_info.next_gc_cutoff()?
-            let next_gc_cutoff = std::cmp::min(gc_info.horizon_cutoff, gc_info.pitr_cutoff);
-
-            let maybe_cutoff = if next_gc_cutoff > timeline.get_ancestor_lsn() {
-                // only include these if they are after branching point; otherwise we would end up
-                // with duplicate updates before the actual branching.
+            // new gc run, which we have no control over. however differently from `Timeline::gc`
+            // we don't consider the `Timeline::disk_consistent_lsn` at all, because we are not
+            // actually removing files.
+            let next_gc_cutoff = cmp::min(gc_info.horizon_cutoff, gc_info.pitr_cutoff);
+
+            // the minimum where we should find the next_gc_cutoff for our calculations.
+            //
+            // next_gc_cutoff in parent branch are not of interest (right now at least), nor do we
+            // want to query any logical size before initdb_lsn.
+            let cutoff_minimum = cmp::max(timeline.get_ancestor_lsn(), timeline.initdb_lsn);
```

Comment:
Looks like this is the only Rust code line that god changes in this block?

Seems not your issue, but I think we should always keep the `timeline.initdb_lsn >= timeline.get_ancestor_lsn()` variant.
I could imagine we can have this violated for broken timelines and timelines being created, but is there any other situation this might happen?

Reply:
Why do we need this here, and not in the GC code? Is it because the GC code also looks at `disk_consistent_lsn`, which cannot be less than initdb_lsn or ancestor_lsn?

Reply:
> Looks like this is the only Rust code line that got changes in this block?

Yes, should had split more. I got too excited about the illustration so combined these :) Also one of the older comments turned out wrong (move the `let next_gc_cutoff = cmp::min(...)` to `GcInfo`).

> Seems not your issue, but I think we should always keep the timeline.initdb_lsn >= timeline.get_ancestor_lsn() variant.

Feels like this is the very first aspects I remember reading from the code, `TimelineMetadata::ancestor_lsn` is not `Option<Lsn>` but `Lsn`, related to `Lsn::is_valid`.

The `timeline.get_ancestor_lsn() == 0` for the default timeline created with new tenant, in `create_empty_timeline`: https://github.com/neondatabase/neon/blob/c38f38dab746ef10b32b350ecd38f5adc4d98a1a/pageserver/src/tenant.rs#L399-L407

> Is it because the GC code also looks at disk_consistent_lsn, which cannot be less than initdb_lsn or ancestor_lsn?

I failed to see this but it would make sense. I haven't checked the "or ancestor_lsn" part.

Reply:
On `GET /v1/tenant/{tenant_id}/timeline` `timeline.get_ancestor_lsn()` gets [turned into None if it's 0](https://github.com/neondatabase/neon/blob/c38f38dab746ef10b32b350ecd38f5adc4d98a1a/pageserver/src/http/routes.rs#L121), and that's what I see for the "main" timelines of my local tenants.

Reply:
Yes, the `Lsn(0)` state is supposed to be for some "initializing" state, which is now `Paused` afaik, and it's essentially the same question as for `Broken` timeline states: how to count their size and should we at all? Maybe not for all errors, but for some?

Not sure if we're ready to answer that in this PR though.

Reply:
Just chatted to confirm that even for active timelines the ancestor_lsn is Lsn(0). I'll resolve this now, try to record the other thread as a TODO.